### PR TITLE
Removes most hard-coded z level checks

### PR DESCRIPTION
**Pull request by Crazylemon64**
I went through the code and made all Z level checks into checks on the level's traits. If a check occurs before the space manager starts up, it will read from the default trait list instead.

Also, I made teleportation checks more consistent - this blocks teleportation on the away mission level, too, instead of the piecemeal prevention that currently exists.

@Fox-McCloud I touched the weather controller a bit, but it should work just the same as before, just working off of level names instead of level indices.

:cl:Crazylemon
fix: Makes teleportation checks more consistent, which will ensure that teleporting in the away mission no longer works.
fix: People on the space ruins level no longer count as dead for assassinate objectives.
/:cl:

### CODE REVIEW

File: code/game/objects/structures/crates_lockers/closets.dm [hunk 1]
```diff
@@ -192,8 +192,7 @@
 			return
 		var/obj/item/weapon/rcs/E = W
 		if(E.rcell && (E.rcell.charge >= E.chargecost))
-			// TODO: Tie into space manager
-			if(!(src.z in config.contact_levels))
+			if(!is_station_contact(src.z))
 				to_chat(user, "<span class='warning'>The rapid-crate-sender can't locate any telepads!</span>")
```
Comment (Crazylemon64) [inline code/game/objects/structures/crates_lockers/closets.dm]: This check...

File: code/game/objects/structures/crates_lockers/crates.dm [hunk 2]
```diff
@@ -84,8 +84,7 @@
 	if(istype(W, /obj/item/weapon/rcs) && !src.opened)
 		var/obj/item/weapon/rcs/E = W
 		if(E.rcell && (E.rcell.charge >= E.chargecost))
-			// TODO: Tie into space manager
-			if(!(src.z in config.player_levels))
+			if(!is_level_reachable(src.z)) // This is inconsistent with the closet sending code
 				to_chat(user, "<span class='warning'>The rapid-crate-sender can't locate any telepads!</span>")
```
Comment (Crazylemon64) [inline code/game/objects/structures/crates_lockers/crates.dm]: Is not the same as this check.

File: _maps/cyberiad.dm [hunk 3]
```diff
@@ -24,7 +24,7 @@ z7 = empty
 
         #define MAP_FILE "cyberiad.dmm"
         #define MAP_NAME "NSS Cyberiad"
-        #define MAP_TRANSITION_CONFIG list(MAIN_STATION = CROSSLINKED, CENTCOMM = SELFLOOPING, TELECOMMS = CROSSLINKED, CONSTRUCTION = CROSSLINKED, MINING = CROSSLINKED, DERELICT = CROSSLINKED, EMPTY_AREA = CROSSLINKED)
+        #define MAP_TRANSITION_CONFIG list(DECLARE_LEVEL(MAIN_STATION,CROSSLINKED,list(STATION_LEVEL,STATION_CONTACT,REACHABLE,AI_OK)), DECLARE_LEVEL(CENTCOMM,SELFLOOPING,list(ADMIN_LEVEL,BLOCK_TELEPORT,IMPEDES_MAGIC)), DECLARE_LEVEL(TELECOMMS,CROSSLINKED,list(REACHABLE,BOOSTS_SIGNAL,AI_OK)), DECLARE_LEVEL(CONSTRUCTION,CROSSLINKED,list(REACHABLE)), DECLARE_LEVEL(MINING,CROSSLINKED,list(REACHABLE,STATION_CONTACT,HAS_WEATHER,ORE_LEVEL,AI_OK)), DECLARE_LEVEL(DERELICT,CROSSLINKED,list(REACHABLE)), DECLARE_LEVEL(EMPTY_AREA,CROSSLINKED,list(REACHABLE)))
```
Comment (tigercat2000) [inline _maps/cyberiad.dm]: please add spacing to this and newlines, it's not even slightly readable like this

File: _maps/metastation.dm [hunk 4]
```diff
@@ -25,7 +25,7 @@ z7 = empty space
 
         #define MAP_FILE "MetaStation.v41A.II.dmm"
         #define MAP_NAME "MetaStation"
-        #define MAP_TRANSITION_CONFIG list(MAIN_STATION = CROSSLINKED, CENTCOMM = SELFLOOPING, TELECOMMS = CROSSLINKED, DERELICT = CROSSLINKED, MINING = CROSSLINKED, EMPTY_AREA = CROSSLINKED, EMPTY_AREA_2 = CROSSLINKED)
+        #define MAP_TRANSITION_CONFIG list(DECLARE_LEVEL(MAIN_STATION,CROSSLINKED,list(STATION_LEVEL,STATION_CONTACT,REACHABLE,AI_OK)), DECLARE_LEVEL(CENTCOMM,SELFLOOPING,list(ADMIN_LEVEL,BLOCK_TELEPORT,IMPEDES_MAGIC)), DECLARE_LEVEL(TELECOMMS,CROSSLINKED,list(REACHABLE,BOOSTS_SIGNAL,AI_OK)), DECLARE_LEVEL(DERELICT,CROSSLINKED,list(REACHABLE)), DECLARE_LEVEL(MINING,CROSSLINKED,list(REACHABLE,STATION_CONTACT,AI_OK)), DECLARE_LEVEL(EMPTY_AREA,CROSSLINKED,list(REACHABLE)), DECLARE_LEVEL(EMPTY_AREA_2,CROSSLINKED,list(REACHABLE)))
```
Comment (tigercat2000) [inline _maps/metastation.dm]: same here

File: _maps/cyberiad.dm [hunk 5]
```diff
@@ -24,7 +24,14 @@ z7 = empty
 
         #define MAP_FILE "cyberiad.dmm"
         #define MAP_NAME "NSS Cyberiad"
-        #define MAP_TRANSITION_CONFIG list(MAIN_STATION = CROSSLINKED, CENTCOMM = SELFLOOPING, TELECOMMS = CROSSLINKED, CONSTRUCTION = CROSSLINKED, MINING = CROSSLINKED, DERELICT = CROSSLINKED, EMPTY_AREA = CROSSLINKED)
+        #define MAP_TRANSITION_CONFIG list(\
```
Comment (tigercat2000) [inline _maps/cyberiad.dm]: proc calls (such as list( )) do not actually need escaped newlines with `\`
you can remove them all with no issue
it only matters in text

Comment (Crazylemon64) [inline _maps/cyberiad.dm]: Isn't this a #define? If I remove this, then the #define won't include any of the other lines.

File: code/controllers/Processes/shuttles.dm [hunk 6]
```diff
@@ -210,8 +209,7 @@ var/const/CALL_SHUTTLE_REASON_LENGTH = 12
 		if(!M.roundstart_move)
 			continue
 		for(var/obj/docking_port/stationary/S in stationary)
-			// TODO: Tie into space manager
-			if(S.z != ZLEVEL_STATION && findtext(S.id, M.id))
+			if(is_station_level(S.z) && findtext(S.id, M.id))
```
Comment (Crazylemon64) [inline code/controllers/Processes/shuttles.dm]: I accidentally negated this condition

File: code/datums/cache/air_alarm.dm [hunk 7]
```diff
@@ -15,8 +15,7 @@ var/global/datum/repository/air_alarm/air_alarm_repository = new()
 		alarms[++alarms.len] = passed_alarm.get_nano_data_console()
 	else
 		for(var/obj/machinery/alarm/alarm in (monitored_alarms ? monitored_alarms : air_alarms))
-			// TODO: Tie into space manager
-			if(!monitored_alarms && alarm.z != ZLEVEL_STATION && alarm.z != ZLEVEL_ASTEROID)
+			if(!monitored_alarms && is_station_contact(alarm.z))
```
Comment (Crazylemon64) [inline code/datums/cache/air_alarm.dm]: I accidentally negated this condition

File: code/datums/diseases/advance/advance.dm [hunk 8]
```diff
@@ -406,8 +406,7 @@ var/list/advance_cures = 	list(
 			AD.Refresh()
 
 		for(var/mob/living/carbon/human/H in shuffle(living_mob_list))
-			// TODO: Tie into space manager
-			if(H.z != ZLEVEL_STATION)
+			if(is_station_level(H.z))
```
Comment (Crazylemon64) [inline code/datums/diseases/advance/advance.dm]: I accidentally negated this condition

File: code/game/machinery/computer/honkputer.dm [hunk 9]
```diff
@@ -69,8 +68,7 @@
 /obj/machinery/computer/HONKputer/attack_hand(var/mob/user as mob)
 	if(..())
 		return
-	// TODO: Tie into space manager
-	if(src.z > ZLEVEL_DERELICT)
+	if(is_away_level(src.z))
 		to_chat(user, "<span class='danger'>Unable to establish a connection</span>: You're too far away from the station!")
```
Comment (Crazylemon64) [inline code/game/machinery/computer/honkputer.dm]: The checks in this file are inconsistent

File: code/game/turfs/space/transit.dm [hunk 10]
```diff
@@ -93,7 +93,8 @@
 	var/max = world.maxx-TRANSITIONEDGE
 	var/min = 1+TRANSITIONEDGE
 
-	var/_z = rand(ZLEVEL_CENTCOMM+1,MAX_Z)	//select a random space zlevel
+	var/datum/space_level/dest = pick(levels_by_trait(REACHABLE))
+	var/_z = dest.zpos	//select a random space zlevel
```
Comment (Crazylemon64) [inline code/game/turfs/space/transit.dm]: Is it OK if being thrown out in transit allows going back to the main station?

File: code/modules/admin/verbs/playsound.dm [hunk 11]
```diff
@@ -81,8 +81,7 @@ var/list/sounds_cache = list()
 
 	for(var/O in global_intercoms)
 		var/obj/item/device/radio/intercom/I = O
-		// TODO: Tie into space manager
-		if(I.z != ZLEVEL_STATION && !ignore_z)
+		if(is_station_level(I.z) && !ignore_z)
```
Comment (Crazylemon64) [inline code/modules/admin/verbs/playsound.dm]: I accidentally negated this condition

File: _maps/cyberiad.dm (hunk 5 above)
Comment (tigercat2000) [inline _maps/cyberiad.dm]: oh shit you're right

File: code/game/turfs/space/transit.dm (hunk 10 above)
Comment (tigercat2000) [inline code/game/turfs/space/transit.dm]: yeah that's fine

**Comment by monster860** [comment]
@Crazylemon64 While you're at it, you should do the Bluespace Wormhole Projector. As it is, nothing prevents you from repeatedly shooting your metabuddy and having him end up in the gamma armory.

**Comment by Crazylemon64** [comment]
@monster860 What do you mean? As in, the gamma armory on z2, where it stays before gamma alert is declared? The wormhole projector only sends people from one portal to the other, afaik.

**Comment by monster860** [comment]
@Crazylemon64 If you shoot a person with it, that person is teleported randomly, regardless of whether they're on z-2 or not.

**Comment by TheDZD** [comment]
Conflicts.

File: code/game/objects/structures/crates_lockers/crates.dm (hunk 2 above)
Comment (Fox-McCloud) [inline code/game/objects/structures/crates_lockers/crates.dm]: the RCS's days are limited, so....

That said, I think this is meant to mostly work in station and mining (for now). I'm not sure the differences between `is_level_reachable` and `is_station_contact` though.

Comment (Crazylemon64) [inline code/game/objects/structures/crates_lockers/crates.dm]: Kind of. There was a config variable called `contact_levels`, which consisted of the asteroid and the main station. That variable has become the `is_station_contact` check we now use. The other is `is_level_reachable`, which as you state, are the levels reachable by space transitions.

Comment (Fox-McCloud) [inline code/game/objects/structures/crates_lockers/crates.dm]: I'd just make them both the same; don't really care which.

**Comment by Crazylemon64** [comment]
@Fox-McCloud I've made the wormhole projector teleport only work on tele-allowed z levels, and made the other checks listed above more consistent.

**Comment by Crazylemon64** [comment]
Whee, more conflicts!

**Comment by Crazylemon64** [comment]
Do you know what time it is...?

**Comment by Crazylemon64** [comment]
say waaaaaa

**Comment by Crazylemon64** [comment]
...

**Comment by Spacemanspark** [comment]
Poor CrazyLemon

**Pull request merged**